runtime/wasm/src/module/mod.rs

```diff
@@ -420,15 +421,19 @@ impl WasmInstance {
             .context("`memory.allocate` function not found")?
             .get1()?;
 
-        let should_interrupt = Arc::new(AtomicBool::new(true));
+        let timeout_stopwatch = Arc::new(std::sync::Mutex::new(TimeoutStopwatch::start_new()));
         if let Some(timeout) = timeout {
             // This task is likely to outlive the instance, which is fine.
             let interrupt_handle = instance.store().interrupt_handle().unwrap();
-            let should_interrupt = should_interrupt.clone();
-            graph::spawn(async move {
-                tokio::time::delay_for(timeout).await;
-                if should_interrupt.load(Ordering::SeqCst) {
-                    interrupt_handle.interrupt()
+            let timeout_stopwatch = timeout_stopwatch.clone();
+            graph::spawn_allow_panic(async move {
+                loop {
+                    let time_left =
+                        timeout.checked_sub(timeout_stopwatch.lock().unwrap().elapsed());
+                    match time_left {
+                        None => break interrupt_handle.interrupt(), // Timed out.
+                        Some(time) => tokio::time::delay_for(time).await,
```

Comment:
Consider having a minimum wait time. There is a diabolical case where the stopwatch pauses with very little time left and this becomes an expensive spin loop that starves other futures. The "right" way to implement this is to awake when the stopwatch is unpaused, but I think you want to expend as little effort as possible and the minimum delay seems simpler.